### ISSUE TITLE
Simulation recreation plan & log

### ISSUE THREAD

**Issue by s-huebler** [boyiguo1/Manuscript-BH_Additive_Cox#19]
Milestones:

- [x] Set up simulation on chpc [Goal date: October 15, Completed date: October 28]
  - [x] Get local simulation working
  - [x] Set up github 
  - [x] Reformat file paths to relative  
    - Had to change the directory name to not have commas (added workaround when linking to scale.res) 
  - [x] Rewrite bash script for job submissions 
  - [x] Restore renv file onto chpc
    -  Added interactive wrapper to rprofile around usethis 
  - [x] Add parallel package dependency to renv file
- [x] Parallelize [Goal date: October 31; Removed from task list November 5]
  - [x] Architecture submission hierarchy 
    - [x] Figure out bottleneck 
      - Bottleneck occurs in one of the models. If the main slow down is cv.glmnet, might need to prioritize set up to allocate cpu access for internal function parallelization. 
    - [x] Consider alternate job submission strategy
      - Need to figure out if running all iterations for one parameter combination on one node is more efficient (pregenerate full data to reduce overhead of multiple jobs, leave room for splitting across models instead)
- [ ] Stabilize BHAM [Goal date: November 30]
  - [x] Update dependencies
  - [x] Change pipes to native
    - Some pipes need to stay in magrittr to keep utility of "." placeholder 
  - [ ] Remove historical functions
    - bacoxph?
    - link.vars?
    - Grouping?
  - [ ]  Rearchitecture/rewrite naming conventions
    - [x] Split helper functions 
      -  construct_smooth_name left as internal under construct_smooth_dat
  - [ ] Add documentation
    - [ ] Roxygen documentation
    - [x] ~In line code explanations~
      - Don't think this is helpful
  - [x] Fully remove BhGLM dependency
    - [x] functions_from_BhGLM file break apart 
      - Not sure if copyright note is binding? Leave for now
      - [x] updates and autoscale moved to updates.r 
        -  update.scale.p.group moved out of bgam function
      - [x] measure.cox changed to measure_cox  and moved to measures.R along with measure.glm
      - [x] Grouping moved to make_group 
    - [x]   Identify any functions not in functions from BhGLM file
      - calculate.cox is used where measure.cox (now measure_cox) needs to be used.
- [ ] Finalize simulation [Goal date: November 30]
  - [ ] Identify issues in code or set up
    - [x] Acosso tuning not working
      - Working intermittently.  Currently skipping with try catch. 
      - Alter? Or just leave and follow mention of issue same as bhglm paper.
    - [x] Cosso model not running 
      - Change to HD function script 
    - [ ] Need time benchmarking
    - [x] Assess feature covariance structure choice
    - [x] Choose evaluation metric  
  - [x] Create scripts to parse results and make visualization
  - [x] Choose number of iterations per parameter combination
    - Choice will be based on journal standards

**Comment by s-huebler** [comment]
Parallelization notes

Current set up:
1 job is 1 iteration. Each job is 1 cpu on 1 node. 
30 parameter combinations, ~100 iterations per combination, 1 instance of data generation and 5 models fit per iteration.
Pros: 
- Clearly top down
- Saves results per iteration (if one instance fails due to random convergence issues, it is possible that other instances that finished would be saved) (able to submit in small batches manually and then pick up later)
Cons:
- Not utilizing full capability of hpc and models are timing out
- Lots of jobs means need to reformat the log/slurm out structure to not mess with directory space

Alternative option 1:
1 job is 1 iteration. Each job is 4 cpu on 1 node. Each cpu fits 1 model OR each cpu is utilized for multithreading during each sequential model fit.
30 parameter combinations, ~100 iterations per combination, 1 instance of data generation and 5 models fit per iteration, 1 model fit per cpu.
Pros: 
- Still top down and still saves results per iteration
- Models to fit in parallel (with multithreading? need to figure out)
- Allows for parallelization either within models or between models (need to know if a single model is bottlenecking before determining which set up would be most beneficial)
Cons:
- Again lots of jobs means need to reformat the log/slurm out structure to not mess with directory space
- If a single model is doing the bottlenecking, and it doesn't already have parallelization capability, would be no reason to use another layer of parallelization
- Not sure about overhead


Alternative option 2:
1 job is 1 parameter combination. Each job is 500/x cpus on x nodes. Each cpu fits 1 model.
30 parameter combinations,  1 instance of data generation per combination (that can be distributed among each iteration), ~100 iterations, 5 models fit per iteration
Pros:
- Cuts down on time spent simulating data (not sure how important)
- Fewer jobs to submit 
Cons:
- Not sure if results can be saved if one model fails.
- Would require full rewrite?

Thoughts.
- Current parallelization not using using all cpu even though 4 available when it should be. 
- cv.glmnet may not be parallelizing properly. 
  - Is it fair to benchmark time for the process run in serial knowing it could be run in parallel?

**Comment by s-huebler** [comment]
Simulation design

- Will need to discuss how many iterations to choose

Evaluation
- Concerns about C-index
  - Hartman N, Kim S, He K, Kalbfleisch JD. Pitfalls of the concordance index for survival outcomes. Stat Med. 2023 Jun 15;42(13):2179-2190. doi: 10.1002/sim.9717. Epub 2023 Mar 28. PMID: 36977424; PMCID: PMC10219847. (https://pmc.ncbi.nlm.nih.gov/articles/PMC10219847/) 
  - Currently being calculated from measure.cox function which fits bacoxph to the model using the identified linear predictors. Need to research more.
  - For now we are just going to continue using C-index

Data generation
- Autoregressive covariance among features. When rho=0 all features are uncorrelated. When rho>0 features are correlated. 
  - Choosing the first 4 features to have true effect artificially amplifies signal. Correlations between first 2 features and any non-significant features are very small, whereas correlations between second 2 features and any non-significant features are moderate. Whereas if we choose 4 randomly in the middle of the feature matrix then it would be more equal. This framework seems problematic because features don't have inherent ordering. 
  - Need to add section to the paper describing why we do this. 
    - Are we just mimicking a real life case in which  linked metabolites are expected to have essentially a group effect? If so, might be better to emulate the feature correlation found in a real life data set. Or to designate groupings among features a priori.
    - Or are we trying to show ability to differentiate/estimate effect of the actual drivers when there is linkage? Because then maybe need to focus more on an evaluation metric that specifically compares this.
    - Or are we trying to show that the correct variables are picked even in situations with linkage? Because then we'd probably need to randomize which feature gets chosen to have an effect so that signal is not amplified. And also in that case we still might want to a priori designate groupings so that we can determine whether models are choosing correct group and also if models are choosing correct feature within a group.

**Comment by s-huebler** [comment]
BhGLM dependencies

- bgam is called in bamlasso for calculating the dispersion only
- bacoxph is not called anywhere

**Comment by s-huebler** [comment]
Documentation notes

bacoxph.R
- bacoxph
  -  Title: Bayesian cox additive model
  - params set and defined in groups, return defined, exported, has DynLib 
- bacoxph.fit
  - Title: Internal fit for bacoxph
  - params set but not defined, keyword internal, not exported
- b.ridge
  - Title: b.ridge
  - params start with ..., keyword internal, not exported

bamlasso.R
- bamlasso
  - Title: Bayesian Spike-and-Slab Lasso Additive Model
  - Params defined, return defined, exported, no examples, has import
- bamlasso_spline.fit
  - Title: Internal spline fit function for bamlasso
  - Params set but not defined, keyword internal, not exported

bgam.R
- bgam
  - Title: Bayesian hierarchical additive models
  - Params defined in mix of group and individual, exported
-  bglm_spline.fit; 
    - Title: Internal spline fit function for bgam; 
    - Params set but not defined, not exported, keyword internal

construct_smooth_data.R
- construct_smooth_data
  - Title: Create a design matrix based on spline configuration
  - Params defined, has description, has imports, has example
- create_smooth_name
  -  Title: Internal name creation for smooth spline
  - not exported, keyword internal

cv_bgam.R
- tune.bgam
- cv.bgam
- generate.foldid
- cv.gam.glm
- cv.gam.lasso
- cv.gam.coxph

family.R
-  NegBin
- Student
- De
- mde
- mt

make_group.R
- make_group
  - Title: Create grouping for spline design matrix
  - Params defined, has imports, exported
- Grouping
  - Title: Grouping function copied from BhGLM

measures.R
- measure_cox
- measure.glm
- link.vars

summarize.R
- parse_var_name
- summarize_ancillary_parameters

updates.R
- update.scale.p.group
- update.prior.sd
- update.ptheta.group
- update.ptheta.network
- update.theta.weights
- auto_scale

variable_selection
- generate_var_table
- bamlasso_var_selection

Single function scripts
- calculate_EDF
  - Title: Calculates the EDF
  - Params set, exported
- plot_smooth_term
  - Title: Plot non-linear functions for BHAM objects
- prepare
- sim_Bai

**Comment by s-huebler** [comment]
# Need help understanding the evaluation of the BHAM model. @boyiguo1 
Each model is evaluated using the measure_cox function which is a custom function from BHAM (adapted from BhGLM) that calculates deviance and c-index. 


# BHAM::measure_cox

It works by fitting a bacoxph model with double exponential (1,0) prior to the Surv object supplied as y and the list of linear predictors supplied as lp. I do not understand this part why the linear predictors is the same row as y.

```
measure_cox <- function(y, lp)
{
  if (NROW(y)!=NROW(lp))
    stop("y and lp should be of the same length", call. = FALSE)
  ff <- bacoxph(y ~ lp, prior=De(1, 0), verbose=FALSE)
  deviance <- -2 * ff$loglik[2]
  cindex <- summary(ff)$concordance[[1]]
  measures <- list(deviance = deviance, Cindex = cindex)
  round(unlist(measures), digits=3)
}

```

# Example of BHAM::measure_cox being used in cosso

As we see here, the process is 
1) model is fit
2) prediction for training and testing data separately which gives lp for each scenario
3) measure_cox function


```
cosso_mdl <-  cosso(x = train_dat |> select(starts_with("X")) |> data.matrix(),
                              y = train_dat |> select(time, status) |> data.matrix(), family = "Cox",
                              nbasis = k, scale = F)

cosso_var <- rep(NA, p) |> `names<-`(names(test_dat |> select(starts_with("X"))))
 cosso_train_lp <- predict.cosso(cosso_mdl,
                                  xnew=train_dat |> select(starts_with("X")) |> data.matrix(),
                                  M=ifelse(!is.null(cosso_tn_mdl), cosso_tn_mdl$OptM, 2), type = "fit")
  cosso_train <- BHAM::measure_cox(Surv(train_dat$time, train_dat$status), cosso_train_lp)

  cosso_test_lp <- predict.cosso(cosso_mdl,
                                 xnew=test_dat |> select(starts_with("X")) |> data.matrix(),
                                 M=ifelse(!is.null(cosso_tn_mdl), cosso_tn_mdl$OptM, 2), type = "fit")
  cosso_test <- BHAM::measure_cox(Surv(test_dat$eventtime, test_dat$status), cosso_test_lp)

```

# BHAM Section of Simulation

Now the BHAM section is different in that lp seems to be stored in the model fit so it is not necessary to predict using the training data. But for testing data, the function called is measure.bh, which is directly from BhGLM and was not translated to BHAM.

```
start_time_bham <- Sys.time()
train_sm_dat <- construct_smooth_data(mgcv_df, train_dat)
train_smooth_data <- train_sm_dat$data

test_sm_dat <- BHAM::make_predict_dat(train_sm_dat$Smooth, dat = test_dat)

bam_group <- BHAM::make_group(names(train_smooth_data))

bamlasso_raw_mdl <- BHAM::bamlasso( x = train_smooth_data, y = Surv(train_dat$time, event = train_dat$status),
                              family = "cox", group = BHAM::make_group(names(train_smooth_data)),
                              ss = c(0.04, 0.5))

blasso_s0_seq <- seq(0.005, 0.1, length.out = 20)    # TODO: need to be optimized
blasso_cv_res <- BHAM::tune.bgam(bamlasso_raw_mdl, nfolds = 5, s0= blasso_s0_seq, verbose = FALSE)

blasso_s0_min <- blasso_cv_res$s0[which.min(blasso_cv_res$deviance)]
bamlasso_mdl <- BHAM::bamlasso( x = train_smooth_data, y = Surv(train_dat$time, event = train_dat$status),
                          family = "cox", group = BHAM::make_group(names(train_smooth_data)),
                          ss = c(blasso_s0_min, 0.5))

bamlasso_train <- BHAM::measure_cox(Surv(train_dat$time, train_dat$status) , bamlasso_mdl$linear.predictors)
#bamlasso_test <- measure.bh(bamlasso_mdl, test_sm_dat, Surv(test_dat$eventtime, test_dat$status))

```

Here is the measure.bh function from BhGLM

```
measure.bh <- function(object, new.x, new.y, new.offset) 
{
  if (missing(new.y)){ 
    y <- object$y
    if (any(class(object) %in% "polr")) y <- object$model[, 1]
  }  
  else y <- new.y
  if (!missing(new.x)) 
    if (NROW(new.x)!=NROW(y))
      stop("'new.x' and 'new.y' should be the same length.", call. = FALSE)
  if (missing(new.offset)) new.offset <- NULL
  if (!is.null(new.offset)) 
    if(NROW(new.offset)!=NROW(y))
      stop("'new.offset' and 'new.y' should be the same length.", call. = FALSE)
  if (!is.null(new.offset) & !missing(new.x) & any(class(object) %in% c("glm", "coxph", "polr"))) 
    new.x <- data.frame(new.x, new.offset)
  
  if (any(class(object) %in% "glm")) 
  {
    mu <- predict(object, newdata=new.x, type="response") 
    if (any(class(object) %in% "negbin")) object$dispersion <- object$theta
    measures <- measure.glm(y, mu, family=object$family$family, dispersion=object$dispersion) 
  }
  
  if (any(class(object) %in% "glmNet") | any(class(object) %in% "bmlasso"))
  {
    family <- object$family
    if (family=="cox")
      if (!is.Surv(y)) stop("'new.y' must be a Surv object")
    if (is.null(object$offset)) object$offset <- FALSE
    else object$offset <- TRUE
    if (missing(new.x)) new.x <- object$x
    newx <- as.matrix(new.x)
    if (family=="cox"){
      lp <- predict(object, newx=newx, newoffset=new.offset)
      lp <- as.vector(lp)
      measures <- measure.cox(y, lp)
    }
    else{ 
      mu <- predict(object, newx=newx, newoffset=new.offset, type="response")
      mu <- as.vector(mu)
      measures <- measure.glm(y, mu, family=family, dispersion=object$dispersion)
    }
  }
  
  if (any(class(object) %in% "coxph")) 
  {
    if (!is.Surv(y)) stop("'new.y' must be a Surv object")
    lp <- predict(object, newdata=new.x)  
    measures <- measure.cox(y, lp)
  }
  
  if (any(class(object) %in% "polr")) 
  {
    if (!is.factor(y)) stop("'new.y' must be a factor")
    probs <- predict(object, newdata=new.x, type="probs")
    measures <- measure.polr(y, probs)
  }
  
  measures    
}

```

# Questions
1) Should I adapt measure.bh from BhGLM into BHAM? If so, I would like to understand what it is doing. Alternatively, should I predict from the model, and if so, what is the best way to do that? 

2) In test_sm_dat, one of the arguments comes from the smooth of train_sm_dat. Need help understanding this one.

**Comment by s-huebler** [comment]
How do we construct confusion matrices (to calculate performance metrics) for spline approximations of true functions? @boyiguo1

**Comment by boyiguo1** [comment]
> How do we construct confusion matrices (to calculate performance metrics) for spline approximations of true functions? [@boyiguo1](https://github.com/boyiguo1)

@s-huebler, As we previous discussed, we can start by following what has been done in https://onlinelibrary.wiley.com/doi/full/10.1002/sim.9483?

What was done in this paper?

**Comment by boyiguo1** [comment]
Questions

>    In test_sm_dat, one of the arguments comes from the smooth of train_sm_dat. Need help understanding this one.

Just curious: Did you try to read the code of `make_predict_dat` either yourself or by chatGPT? Where did you stuck trying to understand.

**Comment by boyiguo1** [comment]
> Should I adapt measure.bh from BhGLM into BHAM? If so, I would like to understand what it is doing. Alternatively, should I predict from the model, and if so, what is the best way to do that?

- I think `measure.bh` should be adapted to BHAM.
- Yes, please understand what it does - from a big picture level
- Yes, you should predict from the model. It is not appropriate to use `predict` function on the trained BHAM objects. Hence, it might require some integration of prediction step from `make_predict_dat, etc, step` to do this.

**Comment by boyiguo1** [comment]
Hey @s-huebler Could we also separate out some of the questions into their own issue/threads. I'm afraid this one issue is too long, and we would forget which questions are being addressed or not. Thanks

**Comment by s-huebler** [comment]
I've separated the questions out into different issues.
